subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelBuildOperationsIntegrationTest.groovy

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal
+
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.integtests.fixtures.timeout.IntegrationTestTimeout
+import org.gradle.test.fixtures.server.http.BlockingHttpServer
+import org.junit.Rule
+import spock.lang.IgnoreIf
+
+import java.time.Instant
+
+@IntegrationTestTimeout(120)
+@IgnoreIf({ GradleContextualExecuter.parallel })
+class WorkerExecutorParallelBuildOperationsIntegrationTest extends AbstractWorkerExecutorIntegrationTest {
+    @Rule
+    BlockingHttpServer blockingHttpServer = new BlockingHttpServer()
+    def buildOperations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def setup() {
+        blockingHttpServer.start()
+        withMultipleActionTaskTypeInBuildScript()
+        buildFile << """
+            task slowTask {
+                doLast { 
+                    ${blockingHttpServer.callFromBuild("slowTask")} 
+                    sleep 10
```

Comment:
Both the `:workTask` and `:slowTask` make http calls, and we assert they are received concurrently. Putting a delay in the `:slowTask` was the only way I could find to ensure that the `:workTask` would finish before the `:slowTask`, in the case that we didn't need to reacquire the project lock.

Reply:
Without this, there's a race to see which task can finish first.

Reply:
One option might be for the "work task should finish first" case that slowTask then blocks waiting for a second task which has a dependency on workTask.  Then we would know that slowTask won't finish before workTask, but if it was broken and workTask was waiting on slowTask, the second block would timeout.  In the other cases we should be able to leave it as is (without the sleep) because slowTask should always finish before workTask.  

This might not be important enough to change unless we see some flakiness.  Arguably, there's still a race condition, but it might not be significant enough to cause an issue.  Your call.

Reply:
There's a [separate test that checks that a dependent task can start while the slowTask is running](https://github.com/gradle/gradle/blob/fbf64fd71ff9924dd26669479ab1b2551c976f4a/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelIntegrationTest.groovy#L539-L573). Note that the dependent task needs to be in a different project for this to work.

So we could delete this test, or leave it and see if flakiness appears. I think I'll add a comment to the test and do the latter.